IGNITE-22712 Describe catalog compaction in README.md (#4325)
apache/ignite-3: modules/catalog-compaction/README.md

# Catalog compaction module

> > **_NOTE_** Compaction has been moved to a separate module to eliminate circular dependencies,

Author: Thanks, fixed.

> as it requires some components that may themselves depend on the catalog module. Please 

Author: Thanks, fixed.

refer to the catalog's module [readme](../catalog/README.md) to see more details about 
catalog service and update log.

## Overview

> During schema changes, the catalog update log stores incremental updates. Each update

Author: Done, thanks

> increases the catalog version. Over time, the log may grow to a humongous  size. To 

Author: thanks, but for me "to a humongous size " sounds better :sunglasses:

> increases the catalog version. Over time, the log may grow to a humongous  size. To 

Author: :thinking:  thanks , but I think it's better to keep original :sunglasses:

address this, snapshotting was introduced to UpdateLog. Snapshotting means replacing 
incremental updates with a snapshot.

> But different components can refer to a specific version of the catalog. Until they 

Author: Different here means something like "distinctive in nature"

finish their work with this version, it cannot be truncated.

This module introduces [CatalogCompactionRunner](src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java)
component. This component handles periodical catalog compaction, ensuring that dropped  versions
of the catalog are no longer needed by any component in the cluster.

## Compaction restrictions

> 1. Catalog compaction can be performed up to the highest version (excluding) which activation time is less 

Reviewer: What does excluding mean here? 
I also do not understand "which activation time..." part.

Author: For example, we have 3 versions of catalog

v1 activation time 1
v2 activation time 3
v3 activation time 5

and we have transaction with activation time 4

"excluding" means that we can trim only v1, because transaction uses catalog v2

Reviewer: Catalog compaction can be performed up to the highest version, while excluding those with activation time lower...

Author: Thank you, I changed it as you suggested.
Sounds better and in my opinion the main meaning does not change.

> 1. Catalog must not be compacted by version which activation time is greater than or equal to earliest

Reviewer 2: You may have few catalog versions v1 and v2 with activation time 1 and 3.
A transaction with activation time of 2 will use catalog of version v1.
So, v1 must not be compacted.

I'd say we can compact catalog up to the highest version (excluding) which activation time is less or equal to the earliest active transaction begin timestamp.

Reviewer 2: Maybe, it is better to replace "the earliest active transaction begin timestamp" with another term like "minimal observable timestamp".
Then say that this time is minimum timestamp from next:
1. tx
2. raft log recover 
and so on.

Author: changed to
```
1. Catalog compaction can be performed up to the highest version (excluding) which activation time is less 
   or equal to the earliest active transaction begin timestamp.
```

   or equal to the earliest active transaction begin timestamp.

> 2. Catalog must not be compacted by version which can be required to replay the raft log during recovery.

Author: Thanks, fixed.

> 3. Index building is stick with a specific catalog version. This version cannot be truncated until

Author: Thanks, fixed.

   the index build is complete.

> 4. Rebalance is stick with a specific catalog version. This version cannot be truncated until the rebalance

Reviewer: Rebalance uses a specific catalog version.

Author: Thanks, fixed.

   is complete.

## Coordinator

> Compaction is performed from single node (compaction coordinator) which is the leader

Author: Thanks, fixed.

> Compaction is performed from single node called compaction coordinator.
> To simplify the selection of the coordinator, it was decided to consider it to be the same node 
> as the leader of the metastorage group.

Reviewer 2: ```suggestion
Compaction is performed from single node (compaction coordinator) which is the leader of the metastorage group for simplicity.
```

Author: Done, thanks

The [ElectionListener](../metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/ElectionListener.java)

> interface was introduced to listen for metastore leader elections.

Author: Thanks, fixed.

## Triggering factors

The process is initiated by one of the following events:

> 1. `low watermark` has been changed (every 5 minutes by default)

Author: Thanks, fixed.

> 1. `low watermark` has been changed

Reviewer 2: Low watermark can be changed too frequently, right?
And the compaction process is not fast/lightweight, right?
If both answers are 'yes', then... is there any back off mechanics?

Author: > Low watermark can be changed too frequently, right?
it's 5 mins by default, but it's configurable :thinking: 

>  is there any back off mechanics?
there is no back off mechanics, only that the process can't start if previous iteration is not finished yet.

Author: Added this into [IGNITE-23059](https://issues.apache.org/jira/browse/IGNITE-23059)

Reviewer 2: Let's just mention `low watermark` is triggered sporadically (every 5min by default).

> 2. the compaction coordinator has been changed

Reviewer: Compaction coordinator changed

Author: Thanks, fixed.

2. the compaction coordinator has been changed

## Overall process description

> When compaction is triggered, two parallel processes are started

Reviewer 2: ```suggestion
When compaction is triggered, two processes are started in parallel
```

Author: Done, thanks

> 1. **Replicas update**. Updates all replication groups with determined minimum begin

Reviewer: **Replicas update**. Updates all replication groups with preset minimum start
   time among all active read-write transactions in the cluster. After some time (see below for details),
   these timestamps are published and become available for the next phase.

Author: Thanks, fixed partially

   time among all active RW transactions in the cluster. After some time (see below for details)
   these timestamps are published and become available for the next phase.

> 2. **Compaction**. Using the timestamps published on the previous stage calculates

Author: Thanks, fixed.

- The second one ("**compaction**") determines the minimum required version of the catalog
  and performs compaction.

![Replicas update](tech-notes/processes.png) 

> **_NOTE_** Even though the compaction process depends on the result of the replicas
update process, they are executed in parallel, so the compaction in the current iteration

> will see the result of the replicas update process obtained in one of the previous iterations.

Reviewer 2: I think this should be merged into previous paragraph.

If my understanding right, 
the first process calculates minimal TX begin time, which will be used later on the next compaction triggering.
the second process uses result (TX begin time) of the previous run of the first process.

Author: Merged to previous paragraph.
```
- The second one ("**compaction**") determines the minimum required version of the catalog
  and performs compaction. At each iteration, it uses the result of the replica update
  process calculated at one of the previous iterations.
```

### Replicas update stage

![Replicas update](tech-notes/replicas-update.png)

This stage consists of the following steps:

1. Each node uses [ActiveLocalTxMinimumBeginTimeProvider](../transactions/src/main/java/org/apache/ignite/internal/tx/ActiveLocalTxMinimumBeginTimeProvider.java)

>    to determine the minimum begin time among all local active RW transactions and sends it to coordinator.

Author: Thanks, fixed.partially

2. Coordinator calculates global minimum and sends it to all nodes using [CatalogCompactionPrepareUpdateTxBeginTimeMessage](src/main/java/org/apache/ignite/internal/catalog/compaction/message/CatalogCompactionPrepareUpdateTxBeginTimeMessage.java).
3. Each node stores this time within replication groups for which the local node is the leader  
   (using [UpdateMinimumActiveTxBeginTimeReplicaRequest](../partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/network/replication/UpdateMinimumActiveTxBeginTimeReplicaRequest.java)
   and [UpdateMinimumActiveTxBeginTimeCommand](../partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/network/command/UpdateMinimumActiveTxBeginTimeCommand.java). 

> 4. This timestamp (let's call it `minTxTime`) is published (become available to compaction process) only

Author: thanks, fixed

>    after checkpoint (flushing partition data to disk).

Reviewer: after checkpoint happens  (partition data is flushed to disk).

Author: thanks, fixed.

### Compaction stage

![Replicas update](tech-notes/compaction.png)

1. Each node determines the local minimum required time, this consists of the following steps:

>    1. Using introduced `RaftGroupStateProvider` determine minimum time among all published

Reviewer: Using the introduced `RaftGroupStateProvider` method to determine the minimum time among all published

Author: thanks, fixed partially

      timestamps (`minTxTime`) in local replication groups.
   2. If `minTxTime` is not published yet, the current iteration of compaction is aborted.

>    3. Select minimum between determined minimum `minTxTime` and current `low watermark`.

Reviewer: The node selects...

Author: thanks, fixed

> 2. Each node sends to the coordinator the calculated local minimum timestamp,

Author: thanks, fixed

>    as well as a set of local replication groups that was used to calculate the local minimum.

Reviewer: as well as to the set of local replication groups that was used to calculate the local minimum.

Author: Didn't get it. what "as well as TO the set" mean here.
We sending not only timestamp but also a set

Reviewer: Then i misunderstood the statement, my bad.

> 3. Coordinator determines global minimum required time.

Author: Thanks, fixed.

4. Using this time, the coordinator determines the version of the catalog up to which (inclusive) history can be trimmed.

> 5. Based on the calculated catalog version, the coordinator determines the list of partitions 

Reviewer: Based on the calculated catalog version, the coordinator determines the list of required partitions

Author: Fixed, thanks

   required by it and compares actual replication groups distribution with what was received
   from remote nodes, and current iteration will be aborted in the following cases:
   1. the logical topology is missing some node required by catalog
   2. some node is missing required replication group
   3. calculated catalog version has an index that is still building
   4. there is an active rebalance, which refer to the calculated (or below calculated) version of the catalog

> 6. Perform catalog compaction.

Reviewer: The node performs catalog compaction.

Author: Fixed, thanks
